x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/ProgressListenableActionFuture.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.index.store.cache;
+
+import org.elasticsearch.action.ActionFuture;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ListenableActionFuture;
+import org.elasticsearch.action.support.AdapterActionFuture;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.collect.Tuple;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+/**
+ * An {@link ActionFuture} that listeners can be attached to. Listeners are executed when the future is completed
+ * or when a given progress is reached. Progression is updated using the {@link #onProgress(long)} method.
+ *
+ * Listeners are executed within the thread that triggers the completion, the failure or the progress update and
+ * the progress value passed to the listeners on execution is the last updated value.
+ */
+class ProgressListenableActionFuture extends AdapterActionFuture<Long, Long> implements ListenableActionFuture<Long> {
+
+    protected final long start;
+    protected final long end;
+
+    // modified under 'this' mutex
+    private volatile List<Tuple<Long, ActionListener<Long>>> listeners;
+    protected volatile @Nullable Long progress; // null if never updated
+    private volatile boolean completed;
+
+    /**
+     * Creates a {@link ProgressListenableActionFuture} that accepts the progression
+     * to be within {@code start} (inclusive) and {@code end} (exclusive) values.
+     *
+     * @param start the start (inclusive)
+     * @param end   the end (exclusive)
+     */
+    ProgressListenableActionFuture(long start, long end) {
+        super();
+        this.start = start;
+        this.end = end;
+        this.completed = false;
+        assert invariant();
+    }
+
+    private boolean invariant() {
+        assert start < end : start + " < " + end;
+        synchronized (this) {
+            assert completed == false || listeners == null;
+            if (progress != null) {
+                assert start <= progress : start + " <= " + progress;
+                assert progress <= end : progress + " <= " + end;
+                assert listeners == null || listeners.stream().allMatch(listener -> progress < listener.v1());
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Updates the progress of the current {@link ActionFuture} with the given value. Calling this method
+     * potentially triggers the execution of one or more listeners that are waiting for the progress
+     * to reach a value lower than the one just updated.
+     *
+     * @param value the new progress value
```

Comment:
I think there might be an off-by-one error lurking here. IMO we should report progress of `value` when the range `⟨start, value⟩` is available, noting that our ranges are inclusive at the start and _exclusive_ at the end. This means I think we can start with `progress == start` (not `null`) indicating that the available range is empty, and require `start < value` below.

Probably a good idea to document this here too.

Reply:
I agree; I've been a bit back and forth here.

Reply:
Should we rename the parameter `progress` to align it with the `{@code progress}` in the paragraph above?